spatialos-sdk/src/worker/schema/bundle.rs

```diff
@@ -0,0 +1,620 @@
+use crate::worker::{component::ComponentId, schema::*, utils::cstr_to_string};
+use spatialos_sdk_sys::worker::*;
+use std::ffi::CString;
+use std::ptr::NonNull;
+
+pub type JsonConversionResult<T> = std::result::Result<(T, Option<String>), String>;
+
+pub struct Bundle {
+    ptr: NonNull<Schema_Bundle>,
+}
+
+impl Bundle {
+    pub fn load(buffer: &[u8]) -> std::result::Result<Self, String> {
+        unsafe {
+            let ptr = Schema_Bundle_Load(buffer.as_ptr(), buffer.len() as u32);
+            let ptr = NonNull::new(ptr)
+                .ok_or_else(|| "Received null pointer from Schema_Bundle_Load".to_string())?;
+            let err = Schema_Bundle_GetError(ptr.as_ptr());
+
+            if !err.is_null() {
+                let msg = cstr_to_string(err);
+                Schema_Bundle_Destroy(ptr.as_ptr());
+                Err(msg)
+            } else {
+                Ok(Bundle { ptr })
+            }
+        }
+    }
+
+    pub fn load_object<T: AsRef<str>, U: AsRef<str>>(
+        &self,
+        qualified_type_name: T,
+        json: U,
+        dest: &mut SchemaObject,
+    ) -> JsonConversionResult<()> {
+        let type_name = CString::new(qualified_type_name.as_ref())
+            .map_err(|_| "Null byte found in 'qualified_type_name'".to_string())?;
+        let json =
+            CString::new(json.as_ref()).map_err(|_| "Null byte found in 'json'".to_string())?;
+
+        unsafe {
+            let success = Schema_Json_LoadObject(
+                self.ptr.as_ptr(),
+                type_name.as_ptr(),
+                json.as_ptr(),
+                dest.as_ptr_mut(),
+            );
+
+            if success == 0 {
+                return Err(Bundle::get_last_error());
+            }
+
+            Ok(((), Bundle::get_last_warning()))
+        }
+    }
+
+    // TODO: Why is this mut?
+    pub fn dump_object<T: AsRef<str>>(
+        &self,
+        qualified_type_name: T,
+        src: &mut SchemaObject,
+    ) -> JsonConversionResult<String> {
```

Comment:
What's this TODO referring to? Is this something that needs to be addressed?

Reply:
It was something for me to investigate. I think its along the same lines of the interior mutability discussion we had a little while ago but need to confirm.